modules/affile/affile-source.c

```diff
@@ -330,95 +331,149 @@ affile_sd_queue(LogPipe *s, LogMessage *msg, const LogPathOptions *path_options,
-                             &self->reader_options,
-                             STATS_LEVEL1,
-                             SCS_FILE,
-                             self->super.super.id,
-                             self->filename->str);
-      /* NOTE: if the file could not be opened, we ignore the last
-       * remembered file position, if the file is created in the future
-       * we're going to read from the start. */
-      
-      log_pipe_append((LogPipe *) self->reader, s);
-      if (!log_pipe_init((LogPipe *) self->reader))
+static gboolean
+_open_file(LogPipe *s, GlobalConfig *cfg)
+{
+  AFFileSourceDriver *self = (AFFileSourceDriver *) s;
+  gint fd;
+
+  if (!affile_sd_open_file(self, self->filename->str, &fd))
+    {
+      if (self->follow_freq == 0)
         {
-          msg_error("Error initializing log_reader, closing fd",
-                    evt_tag_int("fd", fd),
+          msg_error("Error opening file for reading",
+                    evt_tag_str("filename", self->filename->str),
+                    evt_tag_errno(EVT_TAG_OSERROR, errno),
                     NULL);
-          log_pipe_unref((LogPipe *) self->reader);
-          self->reader = NULL;
-          close(fd);
-          return FALSE;
+          return self->super.super.optional;
         }
-      affile_sd_recover_state(s, cfg, proto);
-    }
-  else
-    {
-      msg_error("Error opening file for reading",
-                evt_tag_str("filename", self->filename->str),
-                evt_tag_errno(EVT_TAG_OSERROR, errno),
-                NULL);
-      return self->super.super.optional;
+
+      msg_info("Follow-mode file source not found, deferring open",
+               evt_tag_str("filename", self->filename->str),
+               NULL);
+      fd = -1;
     }
-  return TRUE;
 
+  return _reopen_log_reader(s, cfg, fd);
+}
+
+static gboolean
+affile_sd_init(LogPipe *s)
+{
+  AFFileSourceDriver *self = (AFFileSourceDriver *) s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
+
+  if (!log_src_driver_init_method(s))
+    return FALSE;
+
+  log_reader_options_init(&self->reader_options, cfg, self->super.super.group);
+
+  if (!_check_multiline_options(self))
+    return FALSE;
+
+  self->reader = cfg_persist_config_fetch(cfg, affile_sd_format_persist_name(self));
+  if (self->reader)
+    return _configure_loaded_logreader_from_persist_file(s, cfg);
+
+  return _open_file(s, cfg);
+}
+
+static void
+affile_sd_destroy_reader(gpointer user_data)
+{
+  LogPipe *reader = (LogPipe *) user_data;
+
+  log_pipe_unref(reader);
 }
 
 static gboolean
 affile_sd_deinit(LogPipe *s)
 {
   AFFileSourceDriver *self = (AFFileSourceDriver *) s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
 
   if (self->reader)
     {
       log_pipe_deinit((LogPipe *) self->reader);
-      log_pipe_unref((LogPipe *) self->reader);
+      cfg_persist_config_add(cfg, affile_sd_format_persist_name(self), self->reader, affile_sd_destroy_reader, FALSE);
       self->reader = NULL;
```

Comment:
we could use log_reader_unref() directly as a GDestroyNotify argument, so no need for affile_sd_destroy_reader()